### PR TITLE
ENH: sparse/linalg: support pydata/sparse matrices

### PR DESCRIPTION
**Pull request by pv**
Add support for https://github.com/pydata/sparse sparse matrices to scipy.sparse.linalg.

Although that package is not yet mature, basic support for it in scipy.sparse.linalg appears
to be fairly low-effort.

xref https://github.com/pydata/sparse/issues/293

### CODE REVIEW

File: scipy/sparse/linalg/tests/test_pydata_sparse.py [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+
+
+def test_lobpcg():
+    np.random.seed(1234)
+
+    A_dense = np.random.rand(10, 10)
+    A_dense = A_dense @ A_dense.T
+    A = sparse.COO(A_dense)
+
+    X = np.random.rand(10, 3)
```
Comment (lobpcg) [inline scipy/sparse/linalg/tests/test_pydata_sparse.py]: Please also add here an alternative test with
`X = np.random.randn(10, 1)`
- LOBPCG defaults to the standard 'eig' solver if the size[1] of X is 5 times larger than size[0]

**Comment by rgommers** [comment]
Awesome, thanks a lot for doing this @pv!

@daletovar or @hameerabbasi would be nice if you can find some time to review this with your pydata/sparse hats on.

File: scipy/sparse/linalg/tests/test_pydata_sparse.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+pytestmark = pytest.mark.skipif(sparse is None,
+                                reason="pydata/sparse not installed")
+
+
+def test_isolve_gmres():
+    # Several of the iterative solvers use the same
+    # isolve.utils.make_system wrapper code, so test just one of them.
+    np.random.seed(1234)
+
+    A = sparse.COO(np.random.rand(5, 5))
```
Comment (hameerabbasi) [inline scipy/sparse/linalg/tests/test_pydata_sparse.py]: Would be good to parametrize this over `{DOK, COO, CSD}`  and `xfail` appropriately. It'll expose a lot of bugs/missing features and allow testing.

File: scipy/sparse/linalg/tests/test_pydata_sparse.py [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+    res0 = splin.lsmr(A_dense, b)
+    res = splin.lsmr(A, b)
+    assert_allclose(res[0], res0[0])
+
+
+def test_lsqr():
+    np.random.seed(1234)
+
+    A_dense = np.random.rand(5, 5)
+    A = sparse.COO(A_dense)
```
Comment (hameerabbasi) [inline scipy/sparse/linalg/tests/test_pydata_sparse.py]: Oh, this appears quite a number of times. Maybe a fixture `sparse_class` is in order. 😄

**Review by hameerabbasi — COMMENTED** [review]
LGTM in general, just a couple of comments regarding testing. 😄 It'll allow easy testing for missing features.

File: scipy/sparse/linalg/tests/test_pydata_sparse.py (hunk 2 above)
Comment (pv) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: Is this useful at this stage?

sparse.DOK does not have matvec so nothing will work, and the version currently on pypi does not have CSD.

Comment (hameerabbasi) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: It should hopefully convert to COO when  there’s nothing else available.

Comment (pv) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: Rewrote the test in any case.

File: scipy/sparse/linalg/tests/test_pydata_sparse.py (hunk 3 above)
Comment (pv) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: done

File: scipy/sparse/linalg/tests/test_pydata_sparse.py (hunk 2 above)
Comment (hameerabbasi) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: It’s useful to be able to test upstream in PyData/Sparse and add features related to `xfail`.

Comment (pv) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: Well, marked them with xfail.

DOK does fail with e.g. "'DOK' object has no attribute 'to_scipy_sparse'" and similarly for attribute 'dot'

Comment (daletovar) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: @hameerabbasi do you mean 'GCXS' or is there a reason you're writing 'CSD'?

Comment (hameerabbasi) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: I do mean GCXS... Effects of being tired and sleepless.

Comment (pv) [reply scipy/sparse/linalg/tests/test_pydata_sparse.py]: I think I'll limit this PR to `{COO,DOK}`, which are what is currently available in the pypi version.

**Comment by rgommers** [comment]
The build failure on macOS looks real:
```
lang: scipy/sparse/linalg/dsolve/SuperLU/SRC/zgssvx.c
ar: adding 50 object files to build/temp.macosx-10.6-intel-3.7/libsuperlu_src.a
warning: /Applications/Xcode-9.4.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: archive library: build/temp.macosx-10.6-intel-3.7/libsuperlu_src.a will be fat and ar(1) will not be able to operate on it
ar: adding 50 object files to build/temp.macosx-10.6-intel-3.7/libsuperlu_src.a
ar: build/temp.macosx-10.6-intel-3.7/libsuperlu_src.a is a fat file (use libtool(1) or lipo(1) and ar(1) on it)
ar: build/temp.macosx-10.6-intel-3.7/libsuperlu_src.a: Inappropriate file type or format
...
scipy/sparse/linalg/dsolve/SuperLU/SRC/memory.o" failed with exit status 1
```

Also, I think we want to add `pydata/sparse` as a test dependency here?

**Comment by rgommers** [comment]
This looks great, in it goes. Thanks @pv! And thanks @hameerabbasi and @daletovar for reviewing.

**Pull request merged**